cmake/modules/Threading.cmake

```diff
@@ -0,0 +1,29 @@
+# Get the default threading package for the platform
+function(threading_package_default sdk out_var)
+    if(sdk IN_LIST SWIFT_DARWIN_PLATFORMS)
+      set("${out_var}" "darwin" PARENT_SCOPE)
+    elseif(sdk STREQUAL "LINUX")
+      set("${out_var}" "linux" PARENT_SCOPE)
+    elseif(sdk STREQUAL "WINDOWS")
+      set("${out_var}" "win32" PARENT_SCOPE)
+    elseif(sdk STREQUAL "WASI")
+      set("${out_var}" "none" PARENT_SCOPE)
+    else()
+      set("${out_var}" "pthreads" PARENT_SCOPE)
+    endif()
+endfunction()
+
+# Given the threading package, find the name for the preprocessor
+# define that we need to make.  Also deals with the default platform
+# setting.
+function(threading_package_name sdk out_var)
+  precondition(SWIFT_HOST_VARIANT_SDK)
+  precondition(SWIFT_DARWIN_PLATFORMS)
+
+  string(TOUPPER "${SWIFT_THREADING_PACKAGE}" package)
```

Comment:
I think setting this to `SWIFT_THREADING_PACKAGE` by default is a mistake, as [it is set to the host platform](https://github.com/apple/swift/pull/59287/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR197) by [default](https://github.com/apple/swift/pull/59287/files#diff-5c1eb2d6768a23753402274997290de28b4c44906856c32d858730083df6857eR196). Since it's possible to cross-compile multiple non-host stdlibs, that means it will build all of them with the host threading library name.

I'm not sure how the couple lines just below are supposed to work either, how are you supposed to pass an empty string to `build-script`? I tried various things like `--swift-threading-package=""` and nothing worked. Even if it did, you only have `threading_package_default()` below taking one variable, when it needs two.

This just hit me because I cross-compile the Android armv7 SDK and [it failed](https://github.com/buttaface/swift-android-sdk/runs/6933354497?check_suite_focus=true#step:6:1057) because [it used pthreads](https://github.com/buttaface/swift-android-sdk/runs/6933354497?check_suite_focus=true#step:6:677), which [doesn't support 32-bit here](https://github.com/apple/swift/pull/59287/files#diff-01c32d9dd91f3dff021d98ee0357a5e7c9d0c059eaab416065de0dbe54d2040aR134).

I was confused though because it built fine on the community Android armv7 CI, then I dug into it and it's because the host SDK there is `LINUX` so [it applies that threading library to Android armv7 too](https://ci-external.swift.org/job/oss-swift-RA-linux-ubuntu-16.04-android/lastFailedBuild/console), based on the default logic above. My own Android CI doesn't build for the linux host though, only for Android armv7 [with the flags `--cross-compile-hosts=android-armv7 --skip-local-build`](https://github.com/buttaface/swift-android-sdk/blob/main/.github/workflows/sdks.yml#L131), so the host default logic chooses pthreads for the Android host and fails for armv7 only.

That suggests two fixes:

1. Rethink the way the threading library is chosen by default for the stdlib, so you're not using the host threading library name for all stdlibs built.
2. Add Android to the list of SDKs where a default threading library is chosen, and set a default.

Let me know what you think, @al45tair, you may want to get pthreads working for 32-bit too.

Reply:
Making pthreads work for 32-bit is a fairly trivial fix, I think, and probably worth doing (possibly the right thing is to use `uintptr_t` instead of `uint64_t` for the type of the `swift::once_t`), though maybe we should be picking the `linux` setting on Android?

The `swift-threading-package` setting isn't presently supposed to be explicitly set to empty, which is why that doesn't work. It's there to let you override it and explicitly specify what you want, not the opposite (not sure why you'd need to do that). If you're looking at the headers and have seen the bit in the header file where it tries to guess, that has to be there because of lldb, rather than because I thought that was a good idea.

As for the cross compilation problem, could you file a separate issue for that? The threading code gets used both for the host *and* the target, so I think you're right that the automatic detection might indeed need to be tweaked for the target platform.

Reply:
> maybe we should be picking the linux setting on Android?

I'm not sure: both appear to work on 64-bit platforms, so I will look into it and let you know.

> The swift-threading-package setting isn't presently supposed to be explicitly set to empty, which is why that doesn't work.

I thought [you explicitly included the empty string in this list of threading options](https://github.com/apple/swift/blob/f5bdb858e0e5a6476a00ff20db05cf74a0ded49c/utils/build-script-impl#L2232) to allow people to set it, but I think the bash check a couple lines above wouldn't allow it through anyway.

> It's there to let you override it and explicitly specify what you want, not the opposite (not sure why you'd need to do that)

OK, [this check that I mentioned will never be true then](https://github.com/apple/swift/blob/main/cmake/modules/Threading.cmake#L24)? I thought the intent was for it to signify "auto-detect" for each stdlib, but it currently doesn't do that either.

> As for the cross compilation problem, could you file a separate issue for that?

Will do.

Reply:
PR to fix the remaining 32-bit issues (@compnerd fixed it already for Windows): https://github.com/apple/swift/pull/59586